src/tests/attribute_float_tests.cpp

```diff
@@ -0,0 +1,154 @@
+#include "attribute_float_tests.h"
+
+#include <pbkit/pbkit.h>
+#include <vector>
+
+#include "../test_host.h"
+#include "debug_output.h"
+#include "pbkit_ext.h"
+#include "shaders/precalculated_vertex_shader.h"
+#include "shaders/pixel_shader_program.h"
+
+// clang format off
+static const std::vector<uint32_t> passthrough {
+#include "shaders/precalculated_vertex_shader_4c_texcoords.inl"
+};
+
+static const std::vector<uint32_t> mulByOne {
+#include "shaders/mul_by_one_vertex_shader.inl"
+};
+// clang format on
+
+// Infinities
+static constexpr uint32_t posInf = 0x7F800000;
+static constexpr uint32_t negInf = 0xFF800000;
+// Quiet NaN on x86
+static constexpr uint32_t posNanQ = 0x7FC00000;
+static constexpr uint32_t negNanQ = 0xFFC00000;
+// Signalling NaN on x86
+static constexpr uint32_t posNanS = 0x7F800001;
+static constexpr uint32_t negNanS = 0xFF800001;
+// Max
+static constexpr uint32_t posMax = 0x7F7FFFFF;
+static constexpr uint32_t negMax = 0xFF7FFFFF;
+// Min subnormal
+static constexpr uint32_t posSubNorm= 0x00000001;
+static constexpr uint32_t negSubNorm = 0x80000001;
+
+float f(uint32_t v) {
+    return *((float *)&v);
+}
+
+struct TestConfig {
+  const char *fileName; // must be valid for a filename
+  const char *description;
+  float attribute_value[2];
+};
+
+struct ShaderConfig {
+  const char *shaderName;
+  const std::vector<uint32_t> vertexShader;
+};
+
+static const TestConfig testConfigs[] {
+    { "0_1",  "0 to 1", {  0.f, 1.f }},
+    {"-1_1", "-1 to 1", { -1.f, 1.f }},
+    {"-8_1", "-8 to 1", { -8.f, 1.f }},
+    { "0_8",  "0 to 8", { -1.f, 8.f }},
+    {"-NaNq_NaNq", "-NaN to +NaN (quiet)", {f(negNanQ), f(posNanQ)}},
+    {"-NaNs_NaNs", "-NaN to +NaN (signalling)", {f(negNanS), f(posNanS)}},
+    {"-INF_INF", "-INF to +INF", {f(negInf), f(posInf)}},
+    {"-Max_Max", "-Max to +Max", {f(negMax), f(posMax)}},
+    {"-Min_Min", "-Min to +Min", {f(negSubNorm), f(posSubNorm)}},
+};
+
+static const ShaderConfig shaderConfigs[]{
```

Comment:
Is the intent of the passthrough just to create baseline comparisons to the `mul` case?

If so, maybe instead of doing completely separate images it'd be better to just render two quads with two shaders on the same image?

Reply:
I've updated the test to use both shaders and cover more multipliers